### PR TITLE
TorchTrainDataset now contains a `src` _and_ `dst`.

### PR DESCRIPTION
**Pull request by alxmrs**
Our primary data loader has been modified to differentiate between loading input + boundary vs the label, where the input and label are prognostics. By separating these concepts into two different data sources (well, a data "source" and a data "destination"), we can easily create data loading schedules that allow for training across multiple scales. 

Having "match" or "mix" schedule training examples poses a problem when we have batch sizes greater than 1: we might get mixed resolution examples in a single batch, which will break during the collate step. To address this, this PR includes a batch sampler that we recommend to be used as the default sampler implementation.

### CODE REVIEW

File: src/ocean_emulators/datasets.py [hunk 1]
```diff
@@ -544,14 +586,15 @@ def _get_input_and_label(
...
+        label_all: Float[torch.Tensor, "batch time variable lat lon"],
     ) -> tuple[Input, Prognostic]:
         # grab past steps and prep for model
         total_input = self._prep_tensor_steps(
             prognostic_all[:, : self.hist + 1, :, :, :],
             boundary_all[:, : self.hist + 1, :, :, :],
         )
         # grab future steps, repeat as we do for input
-        label = self._prep_tensor_steps(prognostic_all[:, self.hist + 1 :, :, :, :])
+        label = self._prep_tensor_steps(label_all[:, self.hist + 1 :, :, :, :])
```
Comment (jder) [inline src/ocean_emulators/datasets.py]: This is normalizing the label steps using prognostic_mean/stds. I think probably this is not quite right -- we should use the same means and stds across the whole training run, regardless of whether we're using 1/2 degree or 1 degree as input or output, but probably worth confirming with the broader group. (Perhaps means/stds should be passed in to TorchTrainDataset in that case? Or we should have each datasource carry the training-run-wide means/stds?) Either way we should remove self.label_means/stds if they are unused.

File: src/ocean_emulators/datasets.py [hunk 2]
```diff
@@ -295,25 +295,35 @@ def __getitem__(self, idx):
...
         self.dataset_id: TorchTrainDataset.Id = dataset_id
-        self.raw_data: list[tuple[torch.Tensor, torch.Tensor]] = []
+        self.raw_data: list[tuple[torch.Tensor, torch.Tensor, torch.Tensor]] = []
         self.load_stats: LoadStats | None = None
 
-    def insert(self, all_prognostic: torch.Tensor, all_boundary: torch.Tensor):
-        self.raw_data.append((all_prognostic, all_boundary))
+    def insert(
+        self,
+        all_prognostic: torch.Tensor,
```
Comment (jder) [inline src/ocean_emulators/datasets.py]: I believe the meaning of "prognostic" is "direct output" or maybe "input and output" when we're talking autoregressive. Given that this is no longer used for output, perhaps we should rename it here (and elsewhere in this PR where we've introduced label) to `input`?

File: src/ocean_emulators/datasets.py [hunk 3]
```diff
@@ -382,6 +392,9 @@ def pin_memory(self):
         return self
 
 
+TrainSchedule = Literal["standard", "match", "mix"]
```
Comment (jder) [inline src/ocean_emulators/datasets.py]: I think this should probably be in config.py? It's unused in this file.

File: tests/test_datasets.py [hunk 4]
```diff
@@ -73,11 +102,32 @@ def make_loader(
             cfg.experiment.prognostic_vars_key, cfg.experiment.boundary_vars_key
         )
 
+        match schedule:
+            case "standard":
+                srcs: Iterable[tuple[DataSource, DataSource]] = [(src, src)]
+            case "match":
+                coarsened_src = src.map_data(coarsen_data, suffix="half-size")
```
Comment (jder) [inline tests/test_datasets.py:110]: :sheep: probably we should make `map` operate on the mask as well and use it here rather than this + the next line.

File: tests/test_datasets.py [hunk 5]
```diff
@@ -73,11 +102,32 @@ def make_loader(
...
+                coarsened_src = dataclasses.replace(
+                    coarsened_src, masks=coarsen_masks(src.masks)
+                )
+                scales = [src, coarsened_src]
+                srcs = itertools.permutations(scales, 2)
```
Comment (jder) [inline tests/test_datasets.py]: I think this returns `((src, coarsened_src), (coarsened_src, src))` but I think you want `((src, src), (coarsened_src, coarsened_src))`?

File: tests/test_datasets.py [hunk 6]
```diff
@@ -279,6 +340,91 @@ def test_loader__data_shape(
...
+
+        assert len(samples) == n_samples, (
+            f"Current config {train_config} only supports {n_samples} examples; "
+            f"got {len(samples)}."
+        )
+
+        example_resolutions = []
+        # Only check the first 4 samples; this should be proof enough that everything is
+        # the right shape.
+        for sample in samples[:4]:
```
Comment (jder) [inline tests/test_datasets.py]: Why 4? Won't this be drawn from the first dataset of the loader only? 

I assume this is not passing yet because we don't have the logic where a batch includes only data from one dataset?

**Review by jder — CHANGES_REQUESTED** [review]
Thanks! This looks great; just a few suggestions/questions at this stage.

**Comment by jder** [comment]
@codex review

File: tests/test_datasets.py (hunk 6 above)
Comment (alxmrs) [reply tests/test_datasets.py]: Now, I plan to sample from the whole list, striding by 3. This lets me use the strong assert for the mix schedule.

File: src/ocean_emulators/datasets.py [hunk 7]
```diff
@@ -295,25 +298,36 @@ def __getitem__(self, idx):
...
+        self.raw_data: list[tuple[torch.Tensor, torch.Tensor, torch.Tensor]] = []
         self.load_stats: LoadStats | None = None
 
-    def insert(self, all_prognostic: torch.Tensor, all_boundary: torch.Tensor):
-        self.raw_data.append((all_prognostic, all_boundary))
+    def insert(
+        self,
+        all_input: torch.Tensor,
+        all_boundary: torch.Tensor,
+        all_label: torch.Tensor,
```
Comment (jder) [inline src/ocean_emulators/datasets.py]: :sheep: FWIW the "all_" prefix here was meant to indicate that it was both input and label. Now that we've separated them out I think we can drop it.

File: src/ocean_emulators/datasets.py [hunk 8]
```diff
@@ -523,40 +566,53 @@ def __getitem__(self, idx: int):
...
         return TD
 
     def to_train_data(self, raw_train_data: RawTrainData) -> TrainData:
         train_data = TrainData(self.num_prognostic_channels)
-        for prognostic_all, boundary_all in raw_train_data.raw_data:
-            input, label = self._get_input_and_label(
-                prognostic_all.to(device=self.device, non_blocking=True),
+        for input_all, boundary_all, label_all in raw_train_data.raw_data:
+            input_, label = self._get_input_and_label(
+                input_all.to(device=self.device, non_blocking=True),
```
Comment (jder) [inline src/ocean_emulators/datasets.py]: :sheep: seems a little sad that "get_input_…" takes in something we're calling "input". The former means "the input to the model, including boundary" and the latter means "the input prognostic variables". Maybe it should be "_get_full_input_and_label" or something?

File: src/ocean_emulators/datasets.py [hunk 9]
```diff
@@ -643,6 +699,160 @@ def load_variable_data(var: xr.Variable) -> None:
     wait(futures)
 
 
+class _SimpleSubsetSampler(torch.utils.data.Sampler):
```
Comment (jder) [inline src/ocean_emulators/datasets.py]:
```suggestion
class _SimpleSubsetSampler(Sampler):
```

File: src/ocean_emulators/datasets.py [hunk 10]
```diff
@@ -643,6 +699,160 @@ def load_variable_data(var: xr.Variable) -> None:
...
+
+    def __len__(self):
+        """Calculate total number of batches across all groups."""
+        total_batches = 0
+        for group in self.groups:
+            if self.drop_last:
+                total_batches += len(group) // self.batch_size
+            else:
+                total_batches += (len(group) + self.batch_size - 1) // self.batch_size
+
```
Comment (jder) [inline src/ocean_emulators/datasets.py]: Could also extract the logic to make the sampler from `__iter__` and ask it for its length; not sure if that's better or worse.

File: src/ocean_emulators/datasets.py [hunk 11]
```diff
@@ -643,6 +699,160 @@ def load_variable_data(var: xr.Variable) -> None:
...
+        self.indices = indices
+
+    def __iter__(self):
+        return iter(self.indices)
+
+    def __len__(self):
+        return len(self.indices)
+
+
+class EquivalenceGroupBatchSampler(Sampler[list[int]]):
```
Comment (jder) [inline src/ocean_emulators/datasets.py]: Worth adding some tests directly for this class? Maybe breaking it out to its own file?

File: src/ocean_emulators/datasets.py [hunk 12]
```diff
@@ -643,6 +699,160 @@ def load_variable_data(var: xr.Variable) -> None:
...
+    def __init__(
+        self,
+        dataset_sizes: list[int],
+        batch_size: int,
+        shuffle: bool = True,
+        drop_last: bool = False,
+    ):
+        super().__init__()
+        self.group_size = len(dataset_sizes)
+        self.batch_size = batch_size
```
Comment (jder) [inline src/ocean_emulators/datasets.py]: Might we want different batch sizes for different groups, too? (Don't need to do this now, just thinking out loud.)

File: src/ocean_emulators/datasets.py [hunk 13]
```diff
@@ -643,6 +699,160 @@ def load_variable_data(var: xr.Variable) -> None:
...
+                key = tuple(key) if hasattr(key, "__iter__") else (key,)
+            groups[key].extend(range(cumsum, cumsum + len(ds)))
+            cumsum += len(ds)
+
+        # Convert groups to dataset_sizes format
+        # Sort by key for deterministic ordering across runs
+        sorted_groups = sorted(groups.items(), key=lambda x: str(x[0]))
+
+        # Create instance with computed dataset_sizes
+        instance = cls.__new__(cls)
```
Comment (jder) [inline src/ocean_emulators/datasets.py]: Rather than bypassing the initializer which seems like a smell to me, should `__init__` take a list of ranges instead? Or maybe this from_datasets method should just *be* `__init__`?

File: src/ocean_emulators/datasets.py [hunk 14]
```diff
@@ -643,6 +699,160 @@ def load_variable_data(var: xr.Variable) -> None:
...
+        from collections import defaultdict
+
+        # Group indices by their key
+        groups: dict[tuple, list[int]] = defaultdict(list)
+
+        cumsum = 0
+        for ds in datasets:
+            key = group_key(ds)
+            # Make key hashable if it isn't already
+            if not isinstance(key, (int, str, tuple)):
```
Comment (jder) [inline src/ocean_emulators/datasets.py]: I think we shuld instead enforce that the group key is hashable (eg have the group_key be TorchTrainDataset -> Hashable)

File: src/ocean_emulators/datasets.py [hunk 15]
```diff
@@ -643,6 +699,160 @@ def load_variable_data(var: xr.Variable) -> None:
...
+        dataset_sizes: List of individual dataset sizes. Groups are created based on dataset boundaries,
+            where each dataset forms its own equivalence group.
+        batch_size: Number of samples per batch
+        shuffle: Whether to shuffle indices within groups and shuffle batches globally
+        drop_last: Whether to drop incomplete batches at the end of each group
+    """
+
+    def __init__(
+        self,
+        dataset_sizes: list[int],
```
Comment (jder) [inline src/ocean_emulators/datasets.py]: I suggest we remove the term "Dataset" from here and the description of this class and use "group" instead given these might not be actual individual datasets. (`from_datasets` can then map from actual datasets into abstract groups)

**Review by jder — CHANGES_REQUESTED** [review]
Thanks! Looks great; mostly 🐑 from me and a few minor suggestions.

**Comment by jder** [comment]
@codex review again please

File: src/ocean_emulators/datasets.py [hunk 16]
```diff
@@ -643,6 +699,160 @@ def load_variable_data(var: xr.Variable) -> None:
...
+            # This is regenerated each epoch, giving different orderings
+            all_batches = list(batch_sampler)
+            random.shuffle(all_batches)
+            yield from all_batches
+
+    def __len__(self):
+        """Calculate total number of batches across all groups."""
+        total_batches = 0
+        for group in self.groups:
+            if self.drop_last:
```
[the remark this answers is not in the document]
Comment (jder) [reply src/ocean_emulators/datasets.py]: Indeed. Maybe we should turn on the warning about missing types which would have caught this.

File: src/ocean_emulators/datasets.py (hunk 8 above)
Comment (alxmrs) [reply src/ocean_emulators/datasets.py]: What do you think about using the term "example"? How about `self._to_example` or `self._data_to_example` or `self._make_example`...

File: src/ocean_emulators/datasets.py (hunk 12 above)
Comment (alxmrs) [reply src/ocean_emulators/datasets.py]: Very interesting! I like this idea, but will say YAGNI for now. I'll keep track of it in an issue + comment though.

File: src/ocean_emulators/datasets.py (hunk 10 above)
Comment (alxmrs) [reply src/ocean_emulators/datasets.py]: Sorry, I'm a bit confused by this suggestion. I see that you mean the logic could be extracted into something simpler or more DRY, but I can't quite see how. It's probably because it's been a minute since I looked at this.

Comment (alxmrs) [reply src/ocean_emulators/datasets.py]: I think I get it! Sorry, needed to refresh myself. Making a patch.

File: src/ocean_emulators/datasets.py (hunk 12 above)
Comment (alxmrs) [reply src/ocean_emulators/datasets.py]: I see now! Yes, I refactored the code to do this and it's much simpler.

File: src/ocean_emulators/datasets.py (hunk 13 above)
Comment (alxmrs) [reply src/ocean_emulators/datasets.py]: I agree!

File: src/ocean_emulators/datasets.py (hunk 8 above)
Comment (jder) [reply src/ocean_emulators/datasets.py]: Sounds good to me!

**Comment by jder** [comment]
@alxmrs taking a look now. In the future for big PRs it would be easier for me to track what's going on if you didn't rebase all the commits when making changes. We're going to squash them all anyway so you can just merge main.

**Comment by alxmrs** [comment]
Sure, maybe we can discuss and then document a different development workflow. I'm curious: for big PRs, why a merge over a rebase? Does it have to do with how github trackes comments with commits, for example?

**Comment by jder** [comment]
Yeah, take a look at this page and see if you can tell which commits I should review as a delta from my last review.

**Comment by alxmrs** [comment]
Yeah, github does a poor job at handling this for sure. I don't often check intra PR commit diffs in reviews; mainly I check the comments on the main "timeline" and then look at all files. I'm happy to adjust my workflow if this makes thinks easier to review next time. I may pick your brain later about git habits to adopt -- I think a demo over VC may make it stick for me.

File: src/ocean_emulators/utils/samplers.py [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+        """
+        from collections import defaultdict
+
+        # Group indices by their key
+        groups: dict[Hashable, list[int]] = defaultdict(list)
+
+        cumsum = 0
+        for ds in datasets:
+            key = group_key(ds)
+            assert isinstance(key, (int, str, tuple, Hashable)), (
```
Comment (jder) [inline src/ocean_emulators/utils/samplers.py]:
```suggestion
            assert isinstance(key, Hashable), (
```

I think `int`, `str` and `tuple` of Hashable are all Hashable? Though really we have a type hint here already so hopefully this shouldn't happen.

File: src/ocean_emulators/utils/samplers.py [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+                "`group_key` must be hashable."
+            )
+            groups[key].extend(range(cumsum, cumsum + len(ds)))
+            cumsum += len(ds)
+
+        # Sort by key for deterministic ordering across runs
+        sorted_groups = sorted(groups.items(), key=lambda x: str(x[0]))
```
Comment (jder) [inline src/ocean_emulators/utils/samplers.py]:
```suggestion
        sorted_groups = sorted(groups.items(), key=lambda x: x[0])
```
seems like a possible footgun to me

**Review by jder — APPROVED** [review]
Thanks, looks great! Just two minor (I think) Claudeisms that I don't like :)

**Comment by alxmrs** [comment]
Thanks! Will merge soon.
